src/ut/baseresolver_test.cpp

```diff
@@ -74,127 +71,212 @@ class BaseResolverTest : public ::testing::Test
-  EXPECT_EQ("3.0.0.1:80;transport=TCP",
-            RT(_baseresolver, "").set_host("cpp-common-test.cw-ngv.com").a_resolve());
+  add_white_records(1);
+
+  std::vector<AddrInfo> targets = resolve(1);
+  ASSERT_GT(targets.size(), 0);
+
+  std::string result = ResolverUtils::addrinfo_to_string(resolve(1)[0]);
+  EXPECT_EQ(result, "3.0.0.0:80;transport=TCP");
 }
 
 // Test that IPv4 resolution works when there's multiple correct answers
 TEST_F(BaseResolverTest, IPv4AddressResolutionManyTargets)
 {
-  std::vector<DnsRRecord*> records;
-  records.push_back(ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.1"));
-  records.push_back(ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.2"));
-  records.push_back(ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.3"));
-  records.push_back(ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.4"));
-  records.push_back(ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.5"));
-  records.push_back(ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.6"));
-  _dnsresolver.add_to_cache("cpp-common-test.cw-ngv.com", ns_t_a, records);
-  std::string resolve = RT(_baseresolver, "").set_host("cpp-common-test.cw-ngv.com").a_resolve();
-  EXPECT_THAT(resolve, MatchesRegex("3.0.0.[1-6]:80;transport=TCP"));
+  add_white_records(7);
+
+  std::vector<AddrInfo> targets = resolve(1);
+  ASSERT_GT(targets.size(), 0);
+
+  std::string result = ResolverUtils::addrinfo_to_string(resolve(1)[0]);
+  EXPECT_THAT(result, MatchesRegex("3.0.0.[0-6]:80;transport=TCP"));
 }
 
-// Test that blacklisted A records aren't chosen
-TEST_F(BaseResolverTest, ARecordResolutionWithBlacklist)
+// Test that at least one graylisted record is given out each call, if
+// available.
+TEST_F(BaseResolverTest, ARecordAtLeastOneGray)
 {
-  // Clear the blacklist to start
-  _baseresolver.clear_blacklist();
+  add_white_records(DEFAULT_COUNT);
+  AddrInfo gray_record = ip_to_addr_info("3.0.0.0");
 
-  std::vector<DnsRRecord*> records;
-  records.push_back(ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.1"));
-  records.push_back(ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.2"));
-  _dnsresolver.add_to_cache("cpp-common-test.cw-ngv.com", ns_t_a, records);
-  AddrInfo ai;
-  ai.transport = IPPROTO_TCP;
-  ai.port = 80;
-  DnsRRecord* bl = ResolverUtils::a("cpp-common-test.cw-ngv.com", 3600, "3.0.0.1");
-  ai.address = _baseresolver.to_ip46(bl);
-  _baseresolver.blacklist((const AddrInfo)ai);
+  _baseresolver.blacklist(gray_record);
+  cwtest_advance_time_ms(31000);
+  std::vector<AddrInfo> targets = resolve(DEFAULT_COUNT - 1);
 
-  EXPECT_EQ("3.0.0.2:80;transport=TCP",
-            RT(_baseresolver, "").set_host("cpp-common-test.cw-ngv.com").a_resolve());
-  delete bl;
-  _baseresolver.clear_blacklist();
+  // targets should contain gray
+  EXPECT_TRUE(find(targets.begin(), targets.end(), gray_record) != targets.end());
```

Comment:
I suggested using this idiom in `resolution_contains` but I don't think it's appropriate here. The point was that in `resolution_contains`, the reader has already seen the function name so isn't surprised by the comparison. Here, the reader isn't as prepared, so the code is much harder to parse. Here I would suggest using the `Contains [container matcher](https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md#container-matchers)